### PR TITLE
[Merged by Bors] - feat: extend AddMonoidAlgebra.supDegree API

### PR DESCRIPTION
**Pull request by Ruben-VandeVelde**
This is a part of #7173 that I hope will be uncontroversial.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/MonoidAlgebra/Degree.lean [hunk 1]
```diff
@@ -232,15 +232,84 @@ using the maximum of the exponents. -/
...
+theorem supDegree_add_le {f g : R[A]} :
     (f + g).supDegree D ≤ (f.supDegree D) ⊔ (g.supDegree D) :=
   sup_support_add_le D f g
 
-variable [CovariantClass B B (· + ·) (· ≤ ·)] [CovariantClass B B (Function.swap (· + ·)) (· ≤ ·)]
-  (D : A →+ B) in
-theorem supDegree_mul_le (f g : R[A]) :
-    (f * g).supDegree D ≤ f.supDegree D + g.supDegree D :=
-  sup_support_mul_le (fun {_ _} => (AddMonoidHom.map_add D _ _).le) f g
+theorem supDegree_neg {f : R'[A]} :
```
Comment (adomani) [inline Mathlib/Algebra/MonoidAlgebra/Degree.lean:241]:
```suggestion
@[simp]
theorem supDegree_neg {f : R'[A]} :
```
maybe?

File: Mathlib/Algebra/MonoidAlgebra/Degree.lean [hunk 2]
```diff
@@ -232,15 +232,84 @@ using the maximum of the exponents. -/
...
+    (single a r).supDegree D = D a := by
+  rw [supDegree, Finsupp.support_single_ne_zero a hr, Finset.sup_singleton]
+
+theorem supDegree_single (a : A) (r : R) :
+    (single a r).supDegree D = if r = 0 then ⊥ else D a := by
+  split_ifs with hr <;> simp [supDegree_single_ne_zero, hr]
+
+variable {p q : R[A]}
+
+theorem supDegree_zero : (0 : R[A]).supDegree D = ⊥ := by simp [supDegree]
```
Comment (adomani) [inline Mathlib/Algebra/MonoidAlgebra/Degree.lean:263]:
```suggestion
@[simp]
theorem supDegree_zero : (0 : R[A]).supDegree D = ⊥ := by simp [supDegree]
```
as above!

**Review by adomani — APPROVED** [review]
LGTM!

**Comment by adomani** [comment]
(I did not check whether `simp` already proved the lemmas, though...)

File: Mathlib/Algebra/MonoidAlgebra/Degree.lean [hunk 3]
```diff
@@ -232,15 +232,86 @@ using the maximum of the exponents. -/
...
+  refine' s.induction _ _
+  · rw [Finset.prod_empty, Finset.sum_empty, one_def, supDegree_single]
+    split_ifs; exacts [bot_le, hzero.le]
+  · intro i s his ih
+    rw [Finset.prod_insert his, Finset.sum_insert his]
+    exact (supDegree_mul_le hadd).trans (add_le_add_left ih _)
+
+variable [CovariantClass B B (· + ·) (· < ·)] [CovariantClass B B (Function.swap (· + ·)) (· < ·)]
+
+theorem apply_add_of_supDegree_eq (hD : D.Injective) {ap aq : A}
```
Comment (adomani) [inline Mathlib/Algebra/MonoidAlgebra/Degree.lean:298]: I've definitely seen some form of this lemma.  This probably means that you can use this lemma to golf a similar proof somewhere in the `Polynomial.coeff_mul_of_degree_le`-range (made up name, but I have to go now!).

**Review by jcommelin — APPROVED** [review]
LGTM. Let's merge this. Potential unlocked golfing can be done in a follow-up.

bors merge

**Pull request closed without merging**